src/api/apiTypes.js

```diff
@@ -1,11 +1,234 @@
+  edit_history: MessageEdit[],
+  flags: string[],
+  gravatar_hash: string,
+  id: number,
+  is_me_message: boolean,
+  reactions: SlimEventReaction[],
+  recipient_id: number,
+  sender_email: string,
+  sender_full_name: string,
+  sender_id: number,
+  sender_realm_str: string,
+  sender_short_name: string,
+  stream_id: number, // FixMe: actually only for type `stream`, else absent.
+  subject: string,
+  subject_links: string[],
+  submessages: Message[],
+  timestamp: number,
+  type: 'stream' | 'private',
+};
+
+export type NarrowOperator =
+  | 'is'
+  | 'in'
+  | 'near'
+  | 'id'
+  | 'stream'
+  | 'topic'
+  | 'sender'
+  | 'pm-with'
+  | 'search';
+
+export type NarrowElement = {
+  operand: string,
+  operator?: NarrowOperator, // TODO type: this shouldn't be absent.
+};
+
+export type Narrow = NarrowElement[];
+
+export type RealmEmojiType = {
+  author: {
+    email: string,
+    full_name: string,
+    id: number,
+  },
+  deactivated: boolean,
+  id: number,
+  name: string,
+  source_url: string,
+  // This prevents accidentally using this type as a map.
+  // See https://github.com/facebook/flow/issues/4257#issuecomment-321951793
+  [any]: mixed,
+};
+
+export type RealmEmojiState = {
+  [id: string]: RealmEmojiType,
+};
+
+export type RealmFilter = [string, string, number];
+
+export type Stream = {
+  stream_id: number,
+  description: string,
+  name: string,
+  invite_only: boolean,
+  is_announcement_only: boolean,
+};
+
+export type Subscription = Stream & {
+  color: string,
+  in_home_view: boolean,
+  pin_to_top: boolean,
+  audible_notifications: boolean,
+  desktop_notifications: boolean,
+  email_address: string,
+  push_notifications: boolean,
+  is_old_stream: boolean,
+  push_notifications: boolean,
+  stream_weekly_traffic: number,
+};
+
+export type Topic = {
+  name: string,
+  max_id: number,
+};
+
+/**
+ * A Zulip user.
+ *
+ * For details on the properties, see the Zulip API docs:
+ *   https://zulipchat.com/api/get-all-users#response
+ */
+export type User = {
+  avatar_url: ?string,
+  bot_type?: number,
+  bot_owner?: string,
+  email: string,
+  full_name: string,
+  is_admin: boolean,
+  is_bot: boolean,
+  profile_data?: Object,
+  timezone: string,
+  user_id: number,
+};
+
 export type ApiResponse = {
   result: string,
   msg: string,
 };
 
+export type ApiResponseSuccess = {
+  result: 'success',
+  msg: '',
+};
+
+export type ApiErrorCode =
+  | 'BAD_REQUEST'
+  | 'REQUEST_VARIABLE_MISSING'
+  | 'REQUEST_VARIABLE_INVALID'
+  | 'BAD_IMAGE'
+  | 'REALM_UPLOAD_QUOTA'
+  | 'BAD_NARROW'
+  | 'MISSING_HTTP_EVENT_HEADER'
+  | 'STREAM_DOES_NOT_EXIST'
+  | 'UNAUTHORIZED_PRINCIPAL'
+  | 'UNEXPECTED_WEBHOOK_EVENT_TYPE'
+  | 'BAD_EVENT_QUEUE_ID'
+  | 'CSRF_FAILED'
+  | 'INVITATION_FAILED'
+  | 'INVALID_ZULIP_SERVER';
+
+export type ApiResponseError = {
+  code?: ApiErrorCode,
+  msg: string,
+  result: 'error',
+};
+
+export type ResponseExtractionFunc = (response: Object) => any;
```

Comment:
I appreciate the more specific type here (it was `any`), but given that the rest of this commit is a bulky mechanical change, I'd rather not have it in the same commit.

(This is small enough, and so long as the type-checker still passes the consequences are bounded enough, that it won't stop me from merging the commit as is, though.)